757 mixed prec kernel proc

christophermaynard opened the pull request:
I think this is pretty much ready. I think according to the documentation this should run travis to make sure all is good? Then ready for review!

arporter commented on src/psyclone/parse/kernel.py:
```diff
@@ -611,6 +624,29 @@ def get_kernel_metadata(name, ast):
+
+    :returns 'None' if there is no interface, the interface has no name or has no module procedures
+    '''
+    iname = None
```
I have to admit that I got to this point and then got very confused about what is going on. Please could you tidy up (for pycodestyle) and add comments. I suspect you could extend your new `get_kernel_interface` routine to return the list of kernel names that the interface defines and that might avoid doing this second walk here?

TeranIvy commented:
Sorry to barge into other people's reviews, but could we please be careful about not using `gh_write` access with continuous function spaces such as `w1`? It will be banned in #471 and #138.

christophermaynard commented:
I claim I have made all the required changes (including Iva's). Added a check for multiple interfaces, restructured logic, run pycodestyle and pylint, added tests etc.

arporter commented on src/psyclone/parse/kernel.py:
```diff
@@ -534,40 +534,49 @@ def get_procedure(ast, name, modast):
                             " does not use 'code' as the generic name.".
                             format(name))
-                    bname = statement.bname
+                    bname = statement.bname.lower()
```
On reading the docstring I've realised that this routine returns the parse tree for "the subroutine statement". That's OK when we have a single type-bound procedure but it's not clear when we have multiple implementations referenced from an interface. Please check how this information is used in the caller and update appropriately - you may need to modify how `code` is set below too.
Please could you also the docstring descriptions of the error descriptions to mention interfaces where appropriate.

christophermaynard replied:
Well this is a bit complicated as when the constructor is called on kernelProcedure, it calls get_procedure and the code block is then an internal data member. Its not clear how the code block is used - if at all because all the code generated code seems to be correct and the tests all pass whether I have an interface with several module procedures, or just one type bound procedure. Perhaps we need to discuss this?

christophermaynard replied:
If there is more than one Module procedure in the interface, then it no longer returns the subroutine code block as there would have to be several blocks. Just need to fix up the inlining.

christophermaynard replied:
As discussed, added a check for no code to inlining and a test for this.

arporter commented on src/psyclone/parse/kernel.py:
```diff
@@ -624,29 +633,39 @@ def get_kernel_metadata(name, ast):
+            # count the interfaces, then can be only one!
+            count = count + 1
+            if count >= 2:
+                raise ParseError("Kernel has more than one interface.")
```
Please make this error msg a bit more verbose (e.g. you could pass the kernel name into this routine and use it here) and "...interface; this is forbidden in the LFRic API"

christophermaynard replied:
Add the kernel name as an argument to the procedure and the error message include this and is now more verbose.

arporter commented on src/psyclone/parse/kernel.py:
```diff
@@ -624,29 +633,39 @@ def get_kernel_metadata(name, ast):
-            break
-    return iname
+                iname = statement.name.lower()
+                sname = statement.a.module_procedures
```
You may need to be careful of the `type` of `statement.a.module_procedures` here. Is it a flat list of str?

arporter replied:
`[str(sname) for sname in statement.a.module_procedures]`

christophermaynard replied:
Made the change.

arporter replied:
I can't see any change? (Also, please could you leave the setting of 'resolved' to me so that I know when I'm happy :-) )

arporter commented on src/psyclone/parse/kernel.py:
```diff
@@ -624,29 +633,39 @@ def get_kernel_metadata(name, ast):
+                # procedure name for PSyclone to use.
+            if iname == '':
+                iname = None
+    return iname, sname
```
Please could you convert each element of `sname` to lowercase - that will reduce the need to keep calling `lower()` later on.

christophermaynard replied:
OK, done.

arporter commented on src/psyclone/parse/kernel.py:
```diff
@@ -624,29 +633,39 @@ def get_kernel_metadata(name, ast):
-    :returns: Name 'name' of the parse tree of the interface 
-    :rtype: :str `fparser.one.block_statements.Interface.name`
+    :returns: Name of the parse tree of the interface block and
+              the names of the module procedures. Or None, None if
```
Once you've made the change I've suggested below this needs to say "... procedures (lower case). Or ..."

arporter replied:
Strictly also, the parse tree doesn't have a 'name'. Maybe: "Name of the interface block extracted from the parse tree ..."

christophermaynard replied:
OK. Done.

arporter commented on src/psyclone/parse/kernel.py:
```diff
@@ -624,29 +633,39 @@ def get_kernel_metadata(name, ast):
+              the names of the module procedures. Or None, None if
+              there is no interface or the interface has no nodule procedures.
+    :rtype: : `fparser.one.block_statements.Interface.name`,
+              `fparser.one.block_statements.Interface.a.module_procedures` '''
```
As noted below, I think (unless I've missed something) that the second return value should just be a list of `str`. Possibly the first return value should also just be `str`?

arporter replied:
Please could you also document the error that this routine can raise. Our coding standard also mandates a blank line before the closing "'''" characters of the docstring.

christophermaynard replied:
Done.

arporter commented on src/psyclone/parse/kernel.py:
```diff
@@ -624,29 +633,39 @@ def get_kernel_metadata(name, ast):
+                # If implicit interface (no name) set to none as there is no
+                # procedure name for PSyclone to use.
+            if iname == '':
+                iname = None
```
This line is not covered by the tests (https://codecov.io/gh/stfc/PSyclone/pull/799/diff)

christophermaynard replied:
Restructured code and added a test.

arporter commented on src/psyclone/parse/kernel.py:
```diff
@@ -708,7 +763,7 @@ def __init__(self, ast, name=None):
             # <name/>_type
             found = False
-            for statement, _ in fpapi.walk(ast, -1):
+            for statement, _ in fpapi.walk(ast):
```
Thanks, above and beyond :-)

arporter commented on src/psyclone/tests/parse/kernel_test.py:
```diff
@@ -65,6 +66,52 @@
+    )
+
+
+@pytest.fixture(scope="module", params=[CODE, CODE_INTERFACE])
```
Whoop!

arporter commented on src/psyclone/parse/kernel.py:
```diff
@@ -612,6 +634,39 @@ def get_kernel_metadata(name, ast):
     return ktype
 
 
+def get_kernel_interface(ast):
```
We have a choice here. We can either mandate that (for LFRic Kernels) there must be either a type-bound procedure or a *single* interface, OR, we mandate that if there is an interface then it must be named `<kernel_name>_code`. The latter would allow the Kernel file to contain more than one interface should that ever be needed. If we go for the latter, then this routine will need to be passed the name of the kernel (which I've in fact suggested later for the benefit of the error message). Worth checking what @TeranIvy and @rupertford think?

christophermaynard replied:
I don't believe having multiple interfaces for the kernel is a good idea at all. We won't be doing that in LFRic. That implies the former which the code already does.

arporter commented on src/psyclone/tests/parse/kernel_test.py:
```diff
@@ -133,15 +180,59 @@ def test_getkernelfilepath_caseinsensitive2(tmpdir):
+
+# function get_kernel_interface
+def test_get_kernel_interface_match():
+    ''' Tests that something is returned when searching for an interface'''
```
Please extend: "...for an interface block in a parse tree that contains one."

christophermaynard replied:
Done.

arporter commented on src/psyclone/tests/parse/kernel_test.py:
```diff
@@ -133,15 +180,59 @@ def test_getkernelfilepath_caseinsensitive2(tmpdir):
+    meta1, meta2 = get_kernel_interface(module_parse_tree)
+    assert meta1 is not None
+    assert meta2 is not None
+
```
Please could you add a check that meta1 and meta2 are of the correct type and have the expected values.

christophermaynard replied:
As discussed removed both "match but no check" tests for kernel interface and kernel type-bound procedure as these are kinda pointless.

arporter commented on src/psyclone/tests/parse/kernel_test.py:
```diff
@@ -133,15 +180,59 @@ def test_getkernelfilepath_caseinsensitive2(tmpdir):
+
+
+def test_get_kernel_interface_no_match():
+    ''' Tests for none when searching there is no interface'''
```
Suggest: "...when searching a parse tree that does not contain an interface. "

christophermaynard replied:
Done.

arporter commented on src/psyclone/tests/parse/kernel_test.py:
```diff
@@ -133,15 +180,59 @@ def test_getkernelfilepath_caseinsensitive2(tmpdir):
+
+
+def test_get_kernel_interface_match_correct():
+    ''' Tests that something is returned when searching for an interface'''
```
"... that defines more than one module procedure."

christophermaynard replied:
Done.

arporter commented on src/psyclone/tests/parse/kernel_test.py:
```diff
@@ -133,15 +180,59 @@ def test_getkernelfilepath_caseinsensitive2(tmpdir):
+
+
+def test_get_kernel_interface_double_interface():
+    ''' Tests that parse error occurs for two interfaces.'''
```
"...occurs when the parse tree contains two interfaces." (Subject to agreement on the precise rules we are going to specify for LFRic kernels.)

christophermaynard replied:
Done.

arporter commented on src/psyclone/tests/parse/kernel_test.py:
```diff
@@ -244,15 +334,29 @@ def test_kernelprocedure_notfound():
+
+    '''
+    with pytest.raises(ParseError) as excinfo:
+        my_code = CODE_INTERFACE.replace(
```
Please move this string manipulation to before the `with` block.

christophermaynard commented:
@arporter Perhaps a chat on teams would be a good idea. 
responding to point 1 The precise rules are the same, they haven't changed.
responding to point 2 The whole point is to have multiple module procedures which Fortran run-time picks which one on the arguments. In this case, the number and type is the same (see point 1) but the kind is different - which is currently not controlled by PSyclone and won't be as part of this change.

christophermaynard commented on src/psyclone/parse/kernel.py:
```diff
@@ -534,40 +534,49 @@ def get_procedure(ast, name, modast):
+                                  == subname.lower():
+                        code = statement
+                        break
+                if not code:
```
OK, done.

christophermaynard commented on src/psyclone/parse/kernel.py:
```diff
@@ -534,40 +534,49 @@ def get_procedure(ast, name, modast):
+        # exists for type bound procedure
+        else:
+            code = None
+            for statement, _ in fpapi.walk(modast):
```
Yes, much better.

christophermaynard commented:
@arporter Made changes as required. Please can you review (again). Thanks.

arporter commented on src/psyclone/parse/kernel.py:
```diff
@@ -534,27 +538,44 @@ def get_procedure(ast, name, modast):
-            raise ParseError(
-                "kernel.py:KernelProcedure:get_procedure: Kernel subroutine "
-                "'{0}' not found.".format(bname))
+        # add the name of the tbp to the list of strings to search for.
```
Please could you move this comment into the branch to which it applies (i.e. move to line 559).

christophermaynard replied:
Done.

christophermaynard commented on src/psyclone/parse/kernel.py:
```diff
@@ -634,11 +641,12 @@ def get_kernel_interface(name, ast):
-              there is no interface or the interface has no nodule procedures.
-    :rtype: : `fparser.one.block_statements.Interface.name`,
-              `fparser.one.block_statements.Interface.a.module_procedures` '''
+    :returns: Name of the interface block and the names of the module
```
Done.

christophermaynard commented on src/psyclone/parse/kernel.py:
```diff
@@ -634,11 +641,12 @@ def get_kernel_interface(name, ast):
+    :returns: Name of the interface block and the names of the module
+              procedures (lower case). Or None, None if there is no i
+              nterface or the interface has no nodule procedures.
+    :raises ParseError: if more than one interface is found.
```
Done.

christophermaynard commented on src/psyclone/parse/kernel.py:
```diff
@@ -648,16 +656,19 @@ def get_kernel_interface(name, ast):
+                raise ParseError("Kernel {0} has more than one interface, "
+                                 "this is forbidden in the LFRic API."
+                                 .format(name))
             # Check the interfaces assigns module procedures.
```
Done.

christophermaynard commented on src/psyclone/psyGen.py:
```diff
@@ -32,7 +32,8 @@
 # -----------------------------------------------------------------------------
 # Authors R. W. Ford, A. R. Porter and S. Siso, STFC Daresbury Lab
-# Modified I. Kavcic, Met Office
+# Modified I. Kavcic
```
OK. I guess what I had written was misleading.

christophermaynard commented on src/psyclone/psyGen.py:
```diff
@@ -275,12 +276,19 @@ def gen(self):
         ''' inline all kernel subroutines into the module that are marked for
-            inlining. Avoid inlining the same kernel more than once. '''
+            inlining. Avoid inlining the same kernel more than once.
+            :raises: InternalError if kernel_code is None.
```
Done.

christophermaynard commented on src/psyclone/psyGen.py:
```diff
@@ -275,12 +276,19 @@ def gen(self):
+                    # raise an internal error if kernel_code is None
+                    if kernel._kernel_code is None:
+                        raise InternalError(
+                            "No kernel code to inline for kernel {0}."
```
Done.

christophermaynard commented on src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:
```diff
@@ -32,7 +32,8 @@
 # -----------------------------------------------------------------------------
 # Authors R. W. Ford, A. R. Porter and S. Siso, STFC Daresbury Lab
-# Modified I. Kavcic, Met Office
+# Modified I. Kavcic
```
Done.

christophermaynard commented on src/psyclone/tests/parse/kernel_test.py:
```diff
@@ -106,6 +107,26 @@
 
     )
 
+CODE_DOUBLE_PROCEDURE = (
```
Done.

christophermaynard commented on src/psyclone/tests/parse/kernel_test.py:
```diff
@@ -181,20 +202,11 @@ def test_getkernelfilepath_caseinsensitive2(tmpdir):
-
 def test_get_kernel_interface_no_match():
-    ''' Tests for none when searching there is no interface'''
+    ''' Tests for none when searching there is no interface
```
Yes, that is better.

christophermaynard commented on src/psyclone/tests/parse/kernel_test.py:
```diff
@@ -203,33 +215,50 @@ def test_get_kernel_interface_no_match():
-    module_parse_tree = parse(CODE_INTERFACE.replace(
-        "module procedure sub_code",
-        "module procedure sub_code, more_code"))
+    ''' Tests that the fucntion has correct return when searching for an
```
Done.

christophermaynard commented on src/psyclone/tests/parse/kernel_test.py:
```diff
@@ -203,33 +215,50 @@ def test_get_kernel_interface_no_match():
     kernel_type_name = "interface_procedures"
     meta1, meta2 = get_kernel_interface(kernel_type_name, module_parse_tree)
     assert meta1 == "test_code"
     assert meta2[0] == "sub_code"
```
Done.

christophermaynard commented on src/psyclone/tests/parse/kernel_test.py:
```diff
@@ -203,33 +215,50 @@ def test_get_kernel_interface_no_match():
     assert "Kernel double_interface_kernel has more than one interface, "\
-            "this is forbidden in the LFRic API." \
-            in str(excinfo.value)
+           "this is forbidden in the LFRic API." \
```
Done.

christophermaynard commented:
I have completed changes, including updating the user guide. Back to @arporter for code review. Thanks.

christophermaynard commented:
Oops. Missed a failing test. Now fixed.

arporter commented on src/psyclone/tests/parse/kernel_test.py:
```diff
@@ -65,6 +68,75 @@
+    )
+
+
+@pytest.fixture(scope="module", params=[CODE, CODE_INTERFACE])
```
In order to fix some pylint complaints, please could you add "name='get_code_fragment'" here and then rename the function itself below to e.g. get_code_fragment_fixture.

christophermaynard replied:
done.

christophermaynard commented on doc/user_guide/kernel_layer.rst:
```diff
@@ -106,3 +107,7 @@ procedure.
       procedure ...
       ...
     end type integrate_one_kernel
+
```
OK, done.

christophermaynard commented on src/psyclone/psyGen.py:
```diff
@@ -275,12 +276,20 @@ def gen(self):
         ''' inline all kernel subroutines into the module that are marked for
-            inlining. Avoid inlining the same kernel more than once. '''
+            inlining. Avoid inlining the same kernel more than once.
+            :raises: InternalError if kernel_code (fparser1 AST of kernel) is None.
```
Oops. Done.

christophermaynard commented:
Hi @arporter. I have attempted to update from master. Hopefully it went OK ....

arporter commented on doc/user_guide/kernel_layer.rst:
```diff
@@ -90,24 +90,38 @@ that it allows the metadata to be kept with the code and for it to be
-procedure. If no type-bound procedure is declared, an interface must be
-specified instead as a reference to the kernel code.
+``iterates_over`` and a reference to the kernel code as a type-bound
+procedure.
```
You can do "procedure::" and Sphinx will work its magic - no need to have "::" on a separate line. (Same again below.)

christophermaynard replied:
Thanks, I'll bare that in mind.

arporter commented on doc/user_guide/kernel_layer.rst:
```diff
@@ -90,24 +90,38 @@ that it allows the metadata to be kept with the code and for it to be
+      ... 
+      contains 
+      ... 
+      procedure ... 
```
This one shouldn't have `procedure` in it?

christophermaynard replied:
Oops. Yes. should have noticed that. Thanks.

christophermaynard commented:
Thanks @arporter for helping to get this one through the process.
